### PR TITLE
Verify fixes to limit unlimited pipeline recursion

### PR DESCRIPTION
**Pull request by svanoort**
Pulls in and properly tests https://github.com/cloudbees/groovy-cps/pull/77 -- this is to verify that we're correctly killing bits running on an Executor if the pipeline dies due to excessive  nesting of groovy-cps functions/closures.

Yes, it's a snapshot dep. Yes, it requires building the downstream PR to get it to run correctly because of the special-snowflake nature of the groovy-cps library.  I'm sorry.

### CODE REVIEW

**Comment by jglick** [comment]
````
org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.cloudbees:groovy-cps:jar:1.21-stop-recursion2-SNAPSHOT in azure (https://repo.azure.jenkins.io/public/)
```

Revert the `version` change upstream, `mvn deploy` it, and consume the snapshot here.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java [hunk 1]
```diff
@@ -71,6 +75,78 @@
         assertTrue(exec.isComplete());
     }
 
+    /**
+     * Verify that we kill endlessly recursive CPS code cleanly.
+     */
+    @Test
+    public void failureInNode() throws Exception {
+        String script = "node { assert 1 == 2; } ";
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java]: Uh, why is this an endless recursion? What am I missing here?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java [hunk 2]
```diff
@@ -71,6 +75,78 @@
...
+            for (Executor ex : c.getExecutors()) {
+                if (ex.isBusy()) {
+                    fail(ex.getCurrentExecutable().toString());
+                }
+            }
+        }
+    }
+
+    /**
+     * Verify that we kill endlessly recursive NonCPS code cleanly and don't leave remnants.
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java]: Is this just a native `StackOverflowError`?

Comment (svanoort) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinition2Test.java]: Yes, I wanted a little extra security.  Updated the comments.

**Comment by svanoort** [comment]
Test failure is just windows 8 -- probably a use of sh step somewhere or similar.  Looking into it.

**Comment by svanoort** [comment]
Ugh, just a timeout.

**Comment by svanoort** [comment]
Timeout must point to something actually wrong on Windows with core 2.73.1 -- the timeout in an object.wait call is very puzzling to me, especially since it only occurs in the recursion tests.

**Comment by svanoort** [comment]
Timeout on Windows 8 on core 2.73.1 suggests we've found *something* but no idea what it is. 

> [ERROR] Errors: 
[ERROR] org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition2Test.endlessRecursion(org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition2Test)
[ERROR]   Run 1: CpsFlowDefinition2Test.endlessRecursion:89->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 2: CpsFlowDefinition2Test.endlessRecursion:89->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 3: CpsFlowDefinition2Test.endlessRecursion:89->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 4: CpsFlowDefinition2Test.endlessRecursion:89->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 5: CpsFlowDefinition2Test.endlessRecursion:89->Object.wait:502->Object.wait:-2 � TestTimedOut
[INFO] 
[ERROR] org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition2Test.endlessRecursionNonCPS(org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition2Test)
[ERROR]   Run 1: CpsFlowDefinition2Test.endlessRecursionNonCPS:116->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 2: CpsFlowDefinition2Test.endlessRecursionNonCPS:116->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 3: CpsFlowDefinition2Test.endlessRecursionNonCPS:116->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 4: CpsFlowDefinition2Test.endlessRecursionNonCPS:116->Object.wait:502->Object.wait:-2 � TestTimedOut
[ERROR]   Run 5: CpsFlowDefinition2Test.endlessRecursionNonCPS:116->Object.wait:502->Object.wait:-2 � TestTimedOut

Those lines point to: `WorkflowRun r = jenkins.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());`

**Comment by jglick** [comment]
The thread dump suggests that `SandboxResolvingClassLoader`’s Guava cache is not tolerant to `Error`s. Perhaps. Just a guess. At any rate, looks like a real problem.

**Comment by svanoort** [comment]
Tests pass when I run them locally on a Windows box, including as part of a whole-suite run.   So IDK what's going on here but it seems to be tied to the CI environment.

**Comment by svanoort** [comment]
FFS ci.jenkins.io

**Comment by svanoort** [comment]
@jglick @abayer Would you object to my merging the downstream groovy work + releasing and then releasing this? 

We'd verified (via laborious testing) that the occasional hang is a separate bug with the SandboxResolvingClassloader, which is resolved via https://github.com/jenkinsci/script-security-plugin/pull/160 and does *not* represent a new regression -- the test just happens to be exercising that code a lot more frequently.

**Comment by jglick** [comment]
Well you can `@Ignore` the two failing tests until you have a proper fix. Do not merge a PR with known test failures.

**Comment by svanoort** [comment]
@jglick It's only an intermittent failure, but okay done.

**Pull request merged**